### PR TITLE
add textarea for manual annotations, hide model preds by default, add…

### PR DESCRIPTION
**Pull request by tdhd**
… batch feedback endpoint

### CODE REVIEW

**Comment by tdhd** [comment]
haben jetzt die model predictions per default ausgeblendet, eine textarea (ein keyword pro zeile von den human annotators) und einen `send feedback` button.

die textarea dient dann zum eintragen der konsens-keywords.

habe dafuer einen batch feedback endpoint geadded, im moment passiert beim batch endpoint kein neues training. **muss ich noch einbauen.**

**Comment by tdhd** [comment]
@felixbiessmann was meinst du macht sinn das model training neu anzuschmeissen? ich dachte entweder an was implizites, also nach 10 feedbacks oder mehr oder explizit, dass die GUI einen button bekommt und dann das modell mit training mit schon gelabelten test daten nochmal neu trainiert.

**Comment by felixbiessmann** [comment]
ich bin mir nich sicher, button waer gut, implizit koennte auch gut sein, aber wenn das training zu lange dauert und den arbeitsfluss beim labeln stoert, koennte das vielleicht irritieren. 

man koennte natuerlich auch einfach immer mit dem resultierenden RIS file, das downgeloaded wird, dann beim upload das training neu starten. so muesste man nix am jetzigen design aendern und das retraining wuerde eben immer neu gemacht, wenn neue labels / sortierungen gebraucht werden. 

der nachteil dabei waere, dass wenn alle datenpunkte gelabeled werden, das active learning gar nix bringt. aber wenn man in einer spaeteren version dann was weiss, ueber wie viele labels man braucht, um auf nem test set gute predictions zu kriegen, koennte man dann den labellern sagen ab ner gewissen anzahl von labels (die hoffentlich viel kleiner als das hochgeladene test sample is), dass sie aufhoeren koennen und schon x% der performance haben, die man sich erhoffen kann.

**Comment by tdhd** [comment]
hm ja, wie lange dauert denn auf der instanz die model selection? das ding ist ja, dass beim test upload momentan mit dem aktuellen modell schon vorhersagen für alle artikel generiert werden. besser wäre ja sowas wie predictions on the fly mit dem aktuellen modell, sollte auch möglich sein.
dann ist natürlich keine sortierung so richtig möglich.

**Comment by tdhd** [comment]
ah oder wir machen ändern den prozess so, dass jeweils nur ein artikel, nach aktuellem uncertainty sampling oder so, angezeigt wird und die labeller auf einen knopf drücken um den nächsten zu bekommen (mit predictions). nach einer gewissen anzahl von artikeln wird dann das modell neu trainiert.
was meinst du?

**Comment by felixbiessmann** [comment]
ja, das faend ich gut. also in der jetzigen version isses schon gut so, und der vorschlag mit dem implicit is auch gut so gesehn. 

man muesste halt nur beim trainieren auch diese 'active learning curve' mal berechnen auf nem repraesentativen test set, damit man sagen kann: ab sovielen samples hat das modell x% der performance, die man haette, wenn mans auf allen daten trainiert haette. 

dieses ergebnis kann man dann benutzen um nach soundsovielen gelabeledn beispielen dann vielleicht son popup oder irgendne option fuer den user zu geben: wenn du jetzt das modell wieder trainierst (und anwendest fuer die noch zu labelnden daten), bist du soundso nah an der besten performance (die man mit 100% der neuen labels haette). so waers transparent und der user koennte das selbst entscheiden. 

oder implizit. aber kann verwirren, wenns dann ewig haengt erstma. 

bin mir aber nich sicher, ob bei den kleinen test samples die active learning so nen grossen vorteil bringt. aber das finden wir hoffentlich bald raus.

**Comment by tdhd** [comment]
wegen multi-select nach value counts sortiert: http://davidstutz.github.io/bootstrap-multiselect sieht ganz nett aus (auch aus einen cdn: https://cdnjs.com/libraries/bootstrap-multiselect). gibts auch mit fulltext search zum filtern.

würde das jetzt der einfachheit halber erstmal auf textarea belassen, damit die labeller auch neue keywords hinzufügen können. ist ein bisschen mehr arbeit und fehleranfälliger als so ein multiselect, aber für eine erste version vielleicht ok, oder?

ich würde das jetzt erstmal so mergen mit noch re-training eingebaut.

**Comment by felixbiessmann** [comment]
super! 

ich hab bisschen zeit verloren mit model selection, aber bin grad an der active learning emulation. 
bin ziemlich langsam dabei, die struktur des codes nachzuvollziehn, aber ich krieg hoffentlich noch was fertig heute…

**Comment by tdhd** [comment]
cool, da bin ich echt mal drauf gespannt. ich habe jetzt noch neue features

* journal name full + abbrev.
* publisher name

reingenommen.

dazu hat die HTML seite jetzt neue instructions für die labeller und das modell-retraining kann über einen button gesteuert werden.

**Comment by felixbiessmann** [comment]
ich kaempf grad etwas mit nem development setting, in dem ich daten in classify_cancer packen kann - hast du ne utility function, die ohne den webserver und alles einfach nur die X und y von nem ris file ausgibt?

**Comment by tdhd** [comment]
@felixbiessmann ja, gerade geadded, schau mal hier https://github.com/tdhd/dssg2017/blob/master/django/dkg/classify_cancer_sample.py

musst dafür nur noch https://github.com/tdhd/dssg2017/blob/master/django/dkg/cancer/model_api/model.py#L91 mit 0.66 ersetzen, das mit den settings funktioniert in einem skript nicht.

**Comment by felixbiessmann** [comment]
ah, super, danke. hatte mir jetzt sowas aehnliches gebaut. in der django ipython shell gehn die setting constanten auch.

**Comment by tdhd** [comment]
ah, klar, ja.

habe die aktuelle variante eben auf hetzner online gestellt und trainiere gerade das modell neu dort. mal sehen ob der speicher reicht..

**Comment by tdhd** [comment]
speicher reicht leider nicht und failed schon bei den X encoders.

**Comment by felixbiessmann** [comment]
cool, bin gespannt!

speicher sollte kein problem sein, hab wieder auf sgd umgestellt - oder war das noch nich gemerged. 

ich hab inzwischen auch bemerkt, dass die loss function, die wir anschaun, auch komisch berechnet wird:
wenn die precision ausgerechnet wird, wurden immer die binarized labels benutzt, also die aus clf.predict. 

mir kams dann sinnvoller vor, nen ranking loss zu nehmen - das wuerde irgendwie mehr sinn machen, weils die scores noch mit in account nimmt, unabhaengig vom nem thresholding. 
ich nehm jetzt mal den hier:
http://scikit-learn.org/stable/modules/generated/sklearn.metrics.label_ranking_average_precision_score.html#sklearn.metrics.label_ranking_average_precision_score
fuer die active learning experimente, aber man koennte auch ueberlegen, den loss fuer das modell zu nehmen. macht am ende nicht viel unterschied, glaub ich, aber zumindest fuer die scores waers besser, was zu haben, dass die continuous outputs des models beruecksichtigt. geht aber bei den bisherigen scores nicht so einfach, weil man dann ne sparse matrix und continuous outputs reinsteckt und dafuer sind die nicht definiert in sklearn.

**Comment by felixbiessmann** [comment]
ah, dann stell das mal manuell wieder auf sgd um sorry, dass ich meins noch nich committed hab, aber das is grad noch ziemlich WIP. 

ich nehm jetzt:

text_clf = OneVsRestClassifier(SGDClassifier(loss="log"))
parameters = {'estimator__alpha': (np.logspace(-6,-4,4)).tolist()}

**Comment by tdhd** [comment]
ist ja schon sgd, der batch war nicht commited denke ich.

ich denke swap space wuerde es schon tun, kannst du welchen auf dem hetzner server enabledn? ansonsten kann ich das auch machen, brauche aber sudo von dir.

**Comment by felixbiessmann** [comment]
muesstest jetzt sudo haben - danke!

**Comment by tdhd** [comment]
ah, label ranking AP klingt gut, ja.

hetzner: 4GB swap aktiv, model selection laeuft durch soweit.

das nächste was ich mir anschauen werde, aber nicht mehr heute, sind pipelines fürs training und die inferenz. damit können wir dann auch ohne weiteres andere encoder als hashing vectorizer nehmen für die testdaten.

**Comment by felixbiessmann** [comment]
hatte vergessen zu erwaehnen: 
hab fuer die model selection n_jobs=-1 gesetzt, wenn da mehr als ein core verfuegbar is, wird das die RAM requirements auf der hetzner kiste zu sehr belasten. da koennte man dann n_jobs=1 machen (is default), dann braeuchte man weniger speicher.

**Comment by felixbiessmann** [comment]
Hab gestern noch Ergebnisse bekommen, aber die Experimente brauchen Recht
lange, da ich paar Wiederholungen machen muss, um den noise zu Plätten.

Werd noch plots machen, sobald ich wieder Zeit hab, aber der Code ist ein
gecheckt.

Im grossen und ganzen wie erwartet: bei wenigen labels, also <20%, bringt
Active learning fast immer was, aber der Effekt ist nich so extrem stark.
Liegt auch daran, wie man die Daten aufteilt. Aber zumindest werden die
Plots zeigen, dass es bei wenigen Labels was bringt.

Ausserdem is mir noch eingefallen: wenn wir noch mehr den hashing
vectorizer benutzen. Könnten wir auch schauen, welche Wörter wie wichtig
sind. Und dann der DKG sagen, was sie für Suchbegriffe nehmen sollen, so
als Meta Active learning.

Felix Biessmann <felix.biessmann@gmail.com> schrieb am So., 30. Juli 2017,
23:55:
